iter_archive on zipfiles with better compression type check

Mehdi2402 opened the pull request:
Hello @lhoestq , thank you for your detailed answer on previous PR ! 
I made this new PR because I misused git on the previous one #3347.
Related issue #3272.

# Comments  : 

* For extension check I used the `_get_extraction_protocol` function in **download_manager.py** with a slight change and called it `_get_extraction_protocol_local`: 

**I removed this part :**
 ```python
    elif path.endswith(".tar.gz") or path.endswith(".tgz"):
        raise NotImplementedError(
            f"Extraction protocol for TAR archives like '{urlpath}' is not implemented in streaming mode. Please use `dl_manager.iter_archive` instead."
        )
```
**And also changed :** 
```diff
- extension = path.split(".")[-1]
+ extension = "tar" if path.endswith(".tar.gz") else path.split(".")[-1]
```
The reason for this is a compression like **.tar.gz** will be considered a **.gz** which is handled with **zipfile**, though **tar.gz** can only be opened using **tarfile**.

Please tell me if there's anything to change.

# Tasks : 
- [x] download_manager.py
- [x] streaming_download_manager.py

lhoestq commented on src/datasets/utils/streaming_download_manager.py:
```diff
@@ -512,18 +514,36 @@ def iter_archive(self, urlpath: str):
-                yield (file_path, file_obj)
-                stream.members = []
-            del stream
+        extension = _get_extraction_protocol(urlpath)
```
Currently `_get_extraction_protocol` raises an error for TAR. Therefore you may have to remove the part of `_get_extraction_protocol` that says
```python
    elif path.endswith(".tar.gz") or path.endswith(".tgz"):
        raise NotImplementedError(
            f"Extraction protocol for TAR archives like '{urlpath}' is not implemented in streaming mode. Please use `dl_manager.iter_archive` instead."
        )
```
This verification can be moved in the `_extract` method after calling `_get_extraction_protocol` I think

lhoestq reviewed (commented):
Awesome thank you !

It looks all good to me, I just added a comment to simplify `_get_extraction_protocol_local` a bit since it only has to support local files. There is an error message that also needs to be moved.

To make sure that everything works es expected it would also be nice to have some tests for `iter_archive` in `test_download_manager.py` and `test_streaming_download_manager.py`. I can help you with that if you want

Mehdi2402 commented:
Hello @lhoestq, thank you for your answer.

I don't use pytest a lot so I think I might need some help on it :) but I tried some tests for `streaming_download_manager.py` only. I don't know how to test `download_manager.py` since we need to use local files.

# Comments : 
* In **download_manager.py** I removed some unnecessary imports after the simplification of `_get_extraction_protocol_local`.
* In **streaming_download_manager** I moved the raised Error as suggested.
 
### I also started some tests on `StreamingDownloadManager()` :
* Used an existing zipfile url and added a new one that has a folder and many files : 
```python
TEST_GG_DRIVE_ZIPPED_URL = "https://drive.google.com/uc?export=download&id=1k92sUfpHxKq8PXWRr7Y5aNHXwOCNUmqh"
TEST_GG_DRIVE2_ZIPPED_URL = "https://drive.google.com/uc?export=download&id=1X4jyUBBbShyCRfD-vCO1ZvfqFXP3NEeU"
``` 
* **For now is being tested :**
     * Return type of the function  : should be tuple
     * Files names
     * Files content
 * Added an `xfail` test for the gzip file, because I get a `zipfile.BadZipFile exception`.


 * And lastly, changed the test for `_get_extraction_protocol_throws` since it was moved to `_extract` : 
 ```diff
@pytest.mark.xfail(raises=NotImplementedError)
def test_streaming_dl_manager_get_extraction_protocol_throws(urlpath):
-    _get_extraction_protocol(urlpath)

@pytest.mark.xfail(raises=NotImplementedError)
def test_streaming_dl_manager_get_extraction_protocol_throws(urlpath):
+    StreamingDownloadManager()._extract(urlpath)
```

lhoestq commented on src/datasets/utils/download_manager.py:
```diff
@@ -225,21 +247,37 @@ def iter_archive(self, path):
                     continue
-                file_obj = stream.extractfile(tarinfo)
+                file_obj = io.BufferedReader(zipf.open(member, "r"))
                 yield (file_path, file_obj)
```
Let's raise an error if it's not TAR or ZIP, just in case:
```suggestion
                yield (file_path, file_obj)
            else:
                raise NotImplementedError(f"DownloadManager.iter_archive isn't implemented for '{extension}' file at {path}")
```

lhoestq commented on src/datasets/utils/streaming_download_manager.py:
```diff
@@ -512,18 +515,36 @@ def iter_archive(self, urlpath: str):
+                    file_obj = stream.extractfile(tarinfo)
+                    yield (file_path, file_obj)
+                    stream.members = []
+                del stream
```
```suggestion
                del stream
            else:
                raise NotImplementedError(f"StreamingDownloadManager.iter_archive isn't implemented for '{extension}' file at {path}")
```

lhoestq commented on tests/test_streaming_download_manager.py:
```diff
@@ -39,6 +40,9 @@
 Charmander, fire
 Squirtle, water
 Bulbasaur, grass"""
+TEST_GG_DRIVE2_ZIPPED_URL = "https://drive.google.com/uc?export=download&id=1X4jyUBBbShyCRfD-vCO1ZvfqFXP3NEeU"
+TEST_GG_DRIVE2_FILENAME = ["atoz/test1.txt", "atoz/test2.txt", "atoz/test3/test3.txt"]
+TEST_GG_DRIVE2_CONTENT = ["Erwin is the best character", "this is test 2", "foo"]
```
For such tests I would use another host than Google Drive, which is often unreliable.
In this file you can see some tests that are using files hosted on Google Drive, but this is just to make sure that streaming from Google Drive works (it required a special logic). For the others tests let's use better hosts.

I downloaded your zip and put it here if it can be useful:
https://huggingface.co/datasets/lhoestq/test_zip_txt/resolve/main/atoz.zip

lhoestq commented on tests/test_streaming_download_manager.py:
```diff
@@ -538,3 +542,33 @@ def test_streaming_gg_drive_zipped():
+    "urlpath",
+    [(TEST_GG_DRIVE_GZIPPED_URL)],
+)
+@pytest.mark.xfail()
```
It's ok if `iter_archive` on a gzip file raises `zipfile.BadZipFile` as specified in your test, so I would remove `xfail`

```suggestion
```

lhoestq reviewed (commented):
Cool Thanks ! I think we can use another host that Google Drive for the test file you added.

For the local test you can use the `zip_csv_path` pytest fixture (just add `zip_csv_path` as an argument to the test function). It's a path to a temporary zip file that contains two CSV files `dataset.csv` and `dataset2.csv`. If you're not familiar with pytest I can take care of this

mariosasko commented on src/datasets/utils/download_manager.py:
```diff
@@ -35,6 +37,11 @@
 from .info_utils import get_size_checksum_dict
 from .logging import get_logger
 from .py_utils import NestedDataStructure, map_nested, size_str
+from .streaming_download_manager import (
+    BASE_KNOWN_EXTENSIONS,
+    COMPRESSION_EXTENSION_TO_PROTOCOL,
+    _get_extraction_protocol_with_magic_number,
+)
```
Would move this to `src/datasets/utils/file_utils.py` to avoid importing from `src/datasets/utils/streaming_download_manager.py` to `src/datasets/utils/download_manager.py`

Mehdi2402 replied:
I am not sure if I understood well, do you want me to redefine the function and dictionnaries in `file_utils.py` or just move the imports to `file_utils.py`.
I used the second one in the commit 10cf70078fc34b5b5ba8c6f8afa945163329eb25 Until your response.

Mehdi2402 replied:
> Would move this to `src/datasets/utils/file_utils.py` to avoid importing from `src/datasets/utils/streaming_download_manager.py` to `src/datasets/utils/download_manager.py`

Hello @mariosasko, moving **`COMPRESSION_EXTENSION_TO_PROTOCOL`** to `file_utils.py` creates a cyclic import error as it uses **`COMPRESSION_FILESYSTEMS`** which is imported from `filesystems.py`, and inside `filesystems.py` there are imports from `file_utils.py`.

How do I arrange import which is best for clarity and consistency ?

mariosasko replied:
Maybe let's move `COMPRESSION_EXTENSION_TO_PROTOCOL` to `filesystems.py` then?

mariosasko commented on src/datasets/utils/download_manager.py:
```diff
@@ -64,6 +71,20 @@ class GenerateMode(enum.Enum):
     FORCE_REDOWNLOAD = "force_redownload"
 
 
+def _get_extraction_protocol_local(path: str) -> Optional[str]:
```
A nit for consistency with `streaming_download_manager.py`:
```suggestion
def _get_extraction_protocol(path: str) -> Optional[str]:
```

Mehdi2402 replied:
Hello @mariosasko @lhoestq  Thanks for the answers, I wasn't active lately I will get back to complete this pull request later this week.

lhoestq replied:
thanks @Mehdi2402 !

Mehdi2402 commented:
Hello,
In this Commit was taken into account all the comment escept the `test_download _manager.py`.
I will work on that for the next commit.

Sorry again for being inactive lately in this PR.

Mehdi2402 commented:
> thanks a lot ! This CI seems to have import errors now though ?

Yes sorry about that, it's due to a cyclic import I didn't pay attention to.

Will fix that in the next Commit along with adding the tests to download_manager.

albertvillanova commented:
Hello @Mehdi2402, are you still interested in working on this further?

Mehdi2402 commented:
> Hello @Mehdi2402, are you still interested in working on this further?

Hello @albertvillanova, yes I would like to resume work on this.

albertvillanova commented:
Great, we would like to have this feature.

First, you should fix the conflicts with the main branch, by merging main into your feature branch and then fixing the conflicts by hand. Let us know if you would need some help on this: we can resolve the conflicts for you, so that you can continue your contribution afterwards.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets/pr_3379). All of your documentation changes will be reflected on that endpoint.

mariosasko commented:
I refactored the code to make this PR ready for the final review.

lhoestq commented on src/datasets/download/download_manager.py:
```diff
@@ -69,6 +105,48 @@ def help_message(self):
+            raise NotImplementedError(f"Compression protocol '{compression}' not implemented.")
+
+
+def _get_extraction_protocol(path: str) -> Optional[str]:
```
Maybe all this logic and the streaming one can go in `utils/extract.py` ? If you want you can also keep separate the streaming one and the local one if there are differences

mariosasko replied:
Yes, there are some slight differences, so I think it's best to keep them separated for now for better readability

lhoestq reviewed (approved):
Nice !
